### ISSUE TITLE
Backwards simulation

### ISSUE THREAD

**Issue by davidson16807** [davidson16807/tectonics.js#16]
Follow up on #15, from the same email:

>I am a worldbuilder with a map of a planet which I made with the aid of photshop and fractals which I really like - I have even divided it into rather simplistic plates - but, I would like to be able to take a raster picture of that map, load it into tectonics.js, and run it either forward or backward. Particularly backward, since I know how I want the planet to end up looking... just not how it got that way...

So there's two things going on here:

* **Load map images into simulator**, described in issue #15.

* **use the simulator to work backwards in time**, described here

I never accounted for this in the design of the model and never really intend to. I don't really want this to restrict the design of the model or change how I think about it. I want turnaround times to remain the same.

For implementing this feature, we'll assume the user is fine with a certain loss in predictive power, so any sort of backwards simulation is better than nothing. **If any feature conflicts with backwards simulation, we'll simply have to disable it when running backwards. If we can't disable it, then we'll assume we can no longer support backwards simulation.** This also applies to any new features we implement in the future.

The calculation for plate velocity remains unchanged, as does the integration of velocity into position. The only difference now is **we set our timestep to negative.** 

The same code can be used to create and destroy crust, only now when we create crust, we're actually adding old crust that in a normal run would have subducted into the mantle and been removed from consideration. We don't know the age of the old crust that we add, so we have to guess. Most ocean crust never gets older than 250 million years, so we'll set it to 250 My and say that's as best we can do. 

One consequence of this means age could potentially become negative. Since crust density is a function of age, we could see some really wacky behavior where density becomes negative. Since age never logically becomes negative, we'll have to implement a clamp on age so that this situation never occurs.

Even though we're reusing the create/destroy code, we'll have to create new code to detect where the creation/destruction occurs. Right now the model destroys subducted crust if density exceeds a threshold. When we go backwards though, we only need to destroy where ever an overlap exists between plates. This is a simpler detection scheme, but we still need to decouple the create/destroy code from the detection code before we can implement this. We'll also need a way to switch off between the two detection methods depending on whether timestep is positive or negative. 

Nothing in the upcoming climate model needs to change. Everything in climate is calculated in under a timestep. It's possible we could use results from the previous timestep as an initial guess for climate in the present timestep, but this won't change as timestep becomes negative.

The biggest problem is definitely going to be continental collision. There is no guarantee that continents won't collide when you run backwards, and the model will (rightfully) treat this as a mountain building process. But if you take your backwards-running-simulation and play it forwards, it will appear as if mountains never exist for any particularly good reason. At some point I plan to implement subduction zone congestion, where the plates slow down as their continental crust collides. There's no getting around this - running backwards is going to wreck havok on plans to implement this, so I'll just have to disable it when timestep is negative. 

Erosion is another big sticking point, since it's inherently irreversible. At present, it's represented by something analogous to the heat diffusion equation, which uses the average difference across grid cells (AKA the "Laplacian") to determine whether a cell adds or loses mass. If you apply this in reverse, you will eventually get ridiculous looking, pencil-thin continents that are more like mountain ranges sticking up out of ocean. I'm just going to have to disable this when running backwards.

**Comment by davidson16807** [comment]
So to summarize:

* set a clamp on crust age

* decouple creation/destruction detection from actual creation/destruction

* implement new code for detecting creation/destruction when running backwards, and switch off between the two

* disable erosion when timestep is negative

**Comment by astrographer** [comment]
If backwards sim turns out to be a fairly easy fix(with limitations), I'd say go for it. Otherwise, I'd push it way back in priority.

For a limited form of backwards sim, I'd direct the worldbuilder to gplates.

I'd also suggest the worldbuilder sign up with github. It would be nice to have a larger crowd of kibitzers driving this. Especially kibitzers with more dev skills than I. I don't know how you would regard more devs mucking about in your stuff…

**Comment by davidson16807** [comment]
>I'd also suggest the worldbuilder sign up with github. It would be nice to have a larger crowd of kibitzers driving this. Especially kibitzers with more dev skills than I. 

Definitely agree. 

>I don't know how you would regard more devs mucking about in your stuff…

I would very strongly welcome anyone who wants to contribute code. I'll admit I'm possessive of the model code proper, but there's lots of things I don't care about and just wish they were implemented, particularly file exports and graphics improvements. This is probably the stuff that most people are best equipped to work on, anyways.

**Comment by davidson16807** [comment]
reposting from 16:

>The problem here is that, despite being a physics model, and the physics being deterministic, there are some operations that model the physics that simply cannot be reversed. Even if we could reverse them, that's not to say we wouldn't want to add other subsystems that aren't reversible, and we don't want to limit ourselves in what subsystems we represent (by "subsystem", I mean important things, like fluid simulations for climate, ocean, mantle, etc) so the only way to make this work is to cache the state of the model every few frames for some amount of time, but we're already running up against the limits of javascript as to how much can be represented on typical machines, caching the entire state of the model every few frames will simply multiply what we're doing now. The obvious workaround is to cache at a much lower resolution and not allow the user to play forward from anytime they go back to (lest they get stuck working with the low resolution) This is basically the approach that sim earth took. I agree that it can be done in this form, but this is a lot of work for something that I don't see offers much benefit to the user. I think the user would be better suited having some form of video capture, which is something they can easily get using third party tools.
